purchase_delivery/model/purchase.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    Copyright (C) 2015 Eficent (<http://www.eficent.com/>)
+#               <contact@eficent.com>
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+import time
+from openerp.osv import fields, orm
+from openerp.tools.translate import _
+
+
+class purchase_order(orm.Model):
+    _inherit = 'purchase.order'
+    _columns = {
+        'carrier_id': fields.many2one("delivery.carrier", "Delivery Method",
+                                      help="Complete this field if you plan "
+                                           "to invoice the shipping based on "
+                                           "picking."),
+    }
+
+    def onchange_partner_id(self, cr, uid, ids, part, context=None):
+        result = super(purchase_order, self).onchange_partner_id(
+            cr, uid, ids, part)
+        if part:
+            dtype = self.pool.get('res.partner').browse(
+                cr, uid, part).property_delivery_carrier.id
+            result['value']['carrier_id'] = dtype
+        return result
+
+    def _prepare_order_picking(self, cr, uid, order, context=None):
+        result = super(purchase_order, self)._prepare_order_picking(
+            cr, uid, order, context=context)
+        result.update(carrier_id=order.carrier_id.id)
+        return result
+
+    def delivery_set(self, cr, uid, ids, context=None):
+        line_obj = self.pool.get('purchase.order.line')
+        grid_obj = self.pool.get('delivery.grid')
+        carrier_obj = self.pool.get('delivery.carrier')
+        acc_fp_obj = self.pool.get('account.fiscal.position')
+        for order in self.browse(cr, uid, ids, context=context):
+            src_address_id = order.partner_id
+            dest_address_id = order.dest_address_id or False
+            if (
+                not dest_address_id
+                and order.warehouse_id
```

Comment:
I'm getting this error trying to click the "Add in Quote" button. The warehouse_id field is changed in V8

 File "/home/systemadmin/PycharmProjects/iii-addons/purchase_delivery/model/purchase.py", line 61, in delivery_set
    and order.warehouse_id
AttributeError: 'purchase.order' object has no attribute 'warehouse_id'